File: Sean的leetcode/Leetcode/Basic Calculator.cpp
class Solution {
public:
    int calculate(string s) {
        stack<int> sum_stk;
        stack<int> sign_stk;
        int num = 0, sum = 0;
        int sign = 1; //1: '+', -1: '-'

        for (int i = 0; i < s.size(); i++)
        {
            if (s[i] == ' ')
                continue;

            if (isdigit(s[i]))
                num = num * 10 + s[i] - '0';
            else if (s[i] == '+' || s[i] == '-')
            {
                sum += num * sign;
                num = 0;
                sign = s[i] == '+' ? 1 : -1;
            }
            else if (s[i] == '(')
            {
                sum_stk.push(sum);
                sign_stk.push(sign);
                //reset sum to 0 to calculate sum inside "()"
                sum = 0;
                num = 0;
                sign = 1;
            }
            else if (s[i] == ')')
            {
                sum += num * sign;
                int pre_sum = sum_stk.top();
                int pre_sign = sign_stk.top();
                sum_stk.pop();
                sign_stk.pop();
                pre_sum += sum * pre_sign;
                sum = pre_sum;
                //reset num to 0, ex: (1+2)+3
                num = 0;
                sign = 1;
            }
        }
        sum += num * sign;
        return sum;
    }
};

File: MyLeetCode/python/Basic Calculator.py
###
# Stack
# Time Complexity: O(n)
# Space Complexity: O(n)
###
class Solution(object):
    def calculate(self, s):
        """
        :type s: str
        :rtype: int
        """
        cur = 0
        i = 0
        st = []
        sign = 1
        while i < len(s):
            if s[i].isnumeric():
                l = i
                while i < len(s) and s[i].isnumeric():
                    i += 1
                num = int(s[l:i])
                cur += sign * num
                i -= 1
            elif s[i] == '+' or s[i] == '-':
                sign = 1 if s[i] == '+' else -1         
            elif s[i] == '(':
                st.append(cur)
                st.append(sign)
                cur = 0
                sign = 1
            elif s[i] == ')':
                cur = st.pop() * cur + st.pop()
            i += 1
        return cur
            
        
                                        
                            
                    